pkg/patterns/declarative/metrics.go

```diff
@@ -0,0 +1,342 @@
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative/pkg/manifest"
+)
+
+const DefaultNamespace = "default"
+const ClusterScoped = "_Cluster_"
+
+const (
+	Declarative = "declarative_reconciler"
+)
+
+const (
+	ReconcileCount   = "reconcile_count"
+	ReconcileFailure = "reconcile_failure_count"
+
+	ManagedObjectsRecord = "managed_objects_record"
+)
+
+var (
+	reconcileCount = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: Declarative,
+		Name:      ReconcileCount,
+		Help:      "How many times reconciliation of K8s objects managed by declarative reconciler occurs",
+	}, []string{"group_version_kind", "namespace", "name"})
+
+	reconcileFailure = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: Declarative,
+		Name:      ReconcileFailure,
+		Help:      "How many times reconciliation failure of K8s objects managed by declarative reconciler occurs",
+	}, []string{"group_version_kind", "namespace", "name"})
+
+	managedObjectsRecord = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Subsystem: Declarative,
+		Name:      ManagedObjectsRecord,
+		Help:      "Track the number of objects in manifest",
+	}, []string{"group_version_kind", "namespace", "name"})
+)
+
+func gvkString(gvk schema.GroupVersionKind) string {
+	if len(gvk.Group) == 0 && gvk.Version == "v1" {
+		return gvk.Version + "/" + gvk.Kind
+	}
+
+	return gvk.Group + "/" + gvk.Version + "/" + gvk.Kind
+}
+
+func namespaced(mgr manager.Manager, gvk schema.GroupVersionKind) (bool, error) {
+	mapping, err := mgr.GetRESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return false, err
+	}
+
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+		return true, nil
+	} else {
+		return false, nil
+	}
+}
+
+type reconcileCounter struct {
+	groupVersionKind string
+	namespaced       bool
+	counterVec       *prometheus.CounterVec
+}
+
+func reconcileCounterFor(gvk schema.GroupVersionKind, scope meta.RESTScopeName) reconcileCounter {
+	if scope == meta.RESTScopeNameNamespace {
+		return reconcileCounter{groupVersionKind: gvkString(gvk), namespaced: true, counterVec: reconcileCount}
+	} else {
+		return reconcileCounter{groupVersionKind: gvkString(gvk), namespaced: false, counterVec: reconcileCount}
+	}
+}
+
+func (rc reconcileCounter) reconcileWith(request reconcile.Request) {
+	if rc.namespaced {
+		if len(request.Namespace) == 0 {
+			request.Namespace = DefaultNamespace
```

Comment:
You are right. I'll leave namespace as empty if `reconcile.Request` has empty namespace value. And because of it, I remove code checking whether CRD is namespaced through `RESTMapper`.